xlb/operator/boundary_condition/bc_regularized.py

```diff
@@ -0,0 +1,420 @@
+        self.momentum_flux = SecondMoment()
+
+    # helper function
+    def compute_qi(self):
+        # Qi = cc - cs^2*I
+        dim = self.velocity_set.d
+        Qi = self.velocity_set.cc
+        if dim == 3:
+            diagonal = (0, 3, 5)
+            offdiagonal = (1, 2, 4)
+        elif dim == 2:
+            diagonal = (0, 2)
+            offdiagonal = (1,)
+        else:
+            raise ValueError(f"dim = {dim} not supported")
+
+        # multiply off-diagonal elements by 2 because the Q tensor is symmetric
+        Qi[:, diagonal] += -1.0 / 3.0
+        Qi[:, offdiagonal] *= 2.0
+        return Qi
+
+    @partial(jit, static_argnums=(0,), inline=True)
+    def regularize_fpop(self, fpop, feq):
+        """
+        Regularizes the distribution functions by adding non-equilibrium contributions based on second moments of fpop.
+
+        Parameters
+        ----------
+        fpop : jax.numpy.ndarray
+            The distribution functions.
+        feq : jax.numpy.ndarray
+            The equilibrium distribution functions.
+
+        Returns
+        -------
+        jax.numpy.ndarray
+            The regularized distribution functions.
+        """
+        # Qi = cc - cs^2*I
+        dim = self.velocity_set.d
+        weights = self.velocity_set.w[(slice(None),) + (None,) * dim]
+        # TODO: if I use the following I get NaN ! figure out why!
+        #       Qi = jnp.array(self.compute_qi(), dtype=self.compute_dtype)
+        Qi = jnp.array(self.velocity_set.cc, dtype=self.compute_dtype)
+        if dim == 3:
+            diagonal = (0, 3, 5)
+            offdiagonal = (1, 2, 4)
+        elif dim == 2:
+            diagonal = (0, 2)
+            offdiagonal = (1,)
+        else:
+            raise ValueError(f"dim = {dim} not supported")
+
+        # Qi = cc - cs^2*I
+        # multiply off-diagonal elements by 2 because the Q tensor is symmetric
+        Qi = Qi.at[:, diagonal].add(-1.0 / 3.0)
+        Qi = Qi.at[:, offdiagonal].multiply(2.0)
+
+        # Compute momentum flux of off-equilibrium populations for regularization: Pi^1 = Pi^{neq}
+        f_neq = fpop - feq
+        PiNeq = self.momentum_flux(f_neq)
+        # PiNeq = self.momentum_flux(fpop) - self.momentum_flux(feq)
+
+        # Compute double dot product Qi:Pi1
+        # QiPi1 = np.zeros_like(fpop)
+        # Pi1 = PiNeq
+        QiPi1 = jnp.tensordot(Qi, PiNeq, axes=(1, 0))
+
+        # assign all populations based on eq 45 of Latt et al (2008)
+        # fneq ~ f^1
+        fpop1 = 9.0 / 2.0 * weights * QiPi1
+        fpop_regularized = feq + fpop1
+        return fpop_regularized
+
+    @Operator.register_backend(ComputeBackend.JAX)
+    @partial(jit, static_argnums=(0))
+    def apply_jax(self, f_pre, f_post, boundary_mask, missing_mask):
+        # creat a mask to slice boundary cells
+        boundary = boundary_mask == self.id
+        boundary = jnp.repeat(boundary, self.velocity_set.q, axis=0)
```

Comment:
Can you change jnp.repeat in all BCs to broadcast_in_dim? See equilibirum BC. I am worried that this will create a new array and won't broadcast the values.